pkg/util/ranger/types.go

```diff
@@ -282,6 +282,84 @@ func formatDatum(d types.Datum, isLeftSide bool) string {
 	return fmt.Sprintf("%v", d.GetValue())
 }
 
+// compareLexicographically compares two bounds from two ranges and returns 0, 1, -1
+// for equal, greater than or less than respectively. It gets the two bounds,
+// collations and if each bound is open or closed. In addition, it also gets what type of bound it is.
+// Lower bounds logically can be extended with -infinity and upper bounds can be extended with +infinity.
+func compareLexicographically(tc types.Context, bound1, bound2 []types.Datum, collators []collate.Collator,
+	open1, open2, low1, low2 bool) (int, error) {
```

Comment:
What does `low1` and `low2` mean here?

Reply:
Could you add some examples for this function here, or in the code below?
I think it's hard for one at first sight to grasp the idea behind `open1`, `open2`, `low1`, and `low2`, and how these values lead to different results.

Reply:
The test cases cover these combinations. I do not think it is a good idea to add examples as comments unless it is something simple and not covering combinations like this. We should have comments in the code and I hope I did that. The logic is no complicated. If one value is not equal to the other in the suffix then it is simple. The extra logic for cases where they are equal in the common length and it goes through cases of n1 == n2, n1 > n2 and n1 < n2. For each case, there is additional logic when the range is open or closed.